File: ruby/thousands_separator_test.rb
require_relative 'thousands_separator'
require 'test/unit'

class ThousandsSeparatorTest < Test::Unit::TestCase
  def setup
    @TS_Regexp = ThousandsSeparatorRegExp.new
    @TS_Array = ThousandsSeparatorArray.new
    @TS_Loop =ThousandsSeparatorLoop.new
    @TS_Rec = ThousandsSeparatorRecursive.new

    @numbers = {
      0 => '0',
      12 => '12',
      123 => '123',
      '444' => '444', # test for type string/integer
      1234 => '1,234',
      987654321 => '987,654,321',
      9876543210 => '9,876,543,210'
    } 
  end

  
  # TS_Regexp :: TEST WITH VALUES POSITIVE
  def test_regexp_number_positive
    @numbers.each do |value, expected|
      assert_equal expected, @TS_Regexp.format(value)
    end
  end
  # TS_Regexp :: TEST WITH VALUES NEGATIVE
  def test_regexp_number_negative
    @numbers.each do |value, expected|
      # add auto negative values
      value = value.is_a?(String) ? '-' + value : value * -1
      expected = expected == '0' ? expected : '-' + expected

      assert_equal expected, @TS_Regexp.format(value)
    end
  end

  # TS_Array :: TEST WITH VALUES POSITIVE
  def test_array_number_positive
    @numbers.each do |value, expected|
      assert_equal expected, @TS_Array.format(value)
    end
  end
  # TS_Array :: TEST WITH VALUES NEGATIVE
  def test_array_number_negative
    @numbers.each do |value, expected|
      # add auto negative values
      value = value.is_a?(String) ? '-' + value : value * -1
      expected = expected == '0' ? expected : '-' + expected

      assert_equal expected, @TS_Array.format(value)
    end
  end

  # TS_Loop :: TEST WITH VALUES POSITIVE
  def test_loop_number_positive
    @numbers.each do |value, expected|
      assert_equal expected, @TS_Loop.format(value)
    end
  end
  # TS_Loop :: TEST WITH VALUES NEGATIVE
  def test_loop_number_negative
    @numbers.each do |value, expected|
      # add auto negative values
      value = value.is_a?(String) ? '-' + value : value * -1
      expected = expected == '0' ? expected : '-' + expected

      assert_equal expected, @TS_Loop.format(value)
    end
  end

  # TS_Rec :: TEST WITH VALUES POSITIVE
  def test_rec_number_positive
    @numbers.each do |value, expected|
      assert_equal expected, @TS_Rec.format(value)
    end
  end
  # TS_Rec :: TEST WITH VALUES NEGATIVE
  def test_rec_number_negative
    @numbers.each do |value, expected|
      # add auto negative values
      value = value.is_a?(String) ? '-' + value : value * -1
      expected = expected == '0' ? expected : '-' + expected

      assert_equal expected, @TS_Rec.format(value)
    end
  end
  

  # TEST WITH TYPES & NIL
  def test_string_empty
    assert_equal('0', @TS_Regexp.format(' '))
  end
  def test_string_boolean_true
    assert_equal('0', @TS_Regexp.format(true))
  end
  def test_string_boolean_false
    assert_equal('0', @TS_Regexp.format(false))
  end
  def test_string_nil
    assert_equal('0', @TS_Regexp.format(nil))
  end

  # TEST WITH VALUES
  def test_string_number_text
    assert_equal('0', @TS_Regexp.format('lorem123'))
  end

  def test_string_char_18
    assert_raise(ArgumentError, 'Number lenght is > 15') {
      @TS_Regexp.format('012345678901010101')
    }
  end
end

File: node/thousands_separator.test.js
const {ThousandsSeparatorRegExp, ThousandsSeparatorLoopString, ThousandsSeparatorArray, ThousandsSeparatorRecursive} = require('./thousands_separator.js');

beforeEach(() => {
  TSregexp = new ThousandsSeparatorRegExp;
  TSarray = new ThousandsSeparatorArray;
  TSloop = new ThousandsSeparatorLoopString;
  TSrecursive = new ThousandsSeparatorRecursive;
});

let numbers = {
  0 : '0',
  12 : '12',
  123 : '123',
  '444' : '444', // test for type string/integer
  1234 : '1,234',
  987654321 : '987,654,321',
  9876543210 : '9,876,543,210'
} 

// REGEXP
Object.keys(numbers).forEach(value => {
  let expected = numbers[value]

  test('REGEXP: ' + value + ' => ' + expected, () => {
    expect(TSregexp.format(value)).toBe(expected);
  });
});

// REGEXP - NEGATIVE
Object.keys(numbers).forEach(value => {
  let expected = numbers[value]

  // add auto negative
  if (value !== '0' && value !== 0) { // fix for zero, because in JS (0 * -1 = -0)
    value = (typeof value === 'string') ? '-' + value : -Math.abs(num)
  }
  expected = expected === '0' ? expected : '-' + expected

  test('REGEXP NEGATIVE: ' + value + ' => ' + expected, () => {
    expect(TSregexp.format(value)).toBe(expected);
  });
});


// ARRAY
Object.keys(numbers).forEach(value => {
  let expected = numbers[value]

  test('ARRAY: ' + value + ' => ' + expected, () => {
    expect(TSarray.format(value)).toBe(expected);
  });
});

// ARRAY - NEGATIVE
Object.keys(numbers).forEach(value => {
  let expected = numbers[value]

  // add auto negative
  if (value !== '0' && value !== 0) { // fix for zero, because in JS (0 * -1 = -0)
    value = (typeof value === 'string') ? '-' + value : -Math.abs(num)
  }
  expected = expected === '0' ? expected : '-' + expected

  test('ARRAY NEGATIVE: ' + value + ' => ' + expected, () => {
    expect(TSarray.format(value)).toBe(expected);
  });
});

// LOOP/STRING
Object.keys(numbers).forEach(value => {
  let expected = numbers[value]

  test('LOOP/STRING: ' + value + ' => ' + expected, () => {
    expect(TSloop.format(value)).toBe(expected);
  });
});

// LOOP/STRING
Object.keys(numbers).forEach(value => {
  let expected = numbers[value]

  // add auto negative
  if (value !== '0' && value !== 0) { // fix for zero, because in JS (0 * -1 = -0)
    value = (typeof value === 'string') ? '-' + value : -Math.abs(num)
  }
  expected = expected === '0' ? expected : '-' + expected

  test('LOOP/STRING NEGATIVE: ' + value + ' => ' + expected, () => {
    expect(TSloop.format(value)).toBe(expected);
  });
});

// RECURSIVE
Object.keys(numbers).forEach(value => {
  let expected = numbers[value]

  test('RECURSIVE: ' + value + ' => ' + expected, () => {
    expect(TSarray.format(value)).toBe(expected);
  });
});

// RECURSIVE - NEGATIVE
Object.keys(numbers).forEach(value => {
  let expected = numbers[value]

  // add auto negative
  if (value !== '0' && value !== 0) { // fix for zero, because in JS (0 * -1 = -0)
    value = (typeof value === 'string') ? '-' + value : -Math.abs(num)
  }
  expected = expected === '0' ? expected : '-' + expected

  test('RECURSIVE NEGATIVE: ' + value + ' => ' + expected, () => {
    expect(TSrecursive.format(value)).toBe(expected);
  });
});

// OTHER SEPARATOR
test('REGEXP: Separator = "." ~~> 1234 => return 1.234', () => {
  expect(TSregexp.format('1234', '.')).toBe('1.234');
});

// TEST WITH TYPES & NULL
test('ARRAY: Empty value => return 0', () => {
  expect(TSarray.format(' ')).toBe('0');
});
test('ARRAY: Boolean true => return 0', () => {
  expect(TSarray.format(true)).toBe('0');
});
test('ARRAY: Boolean false => return 0', () => {
  expect(TSarray.format(false)).toBe('0');
});
test('ARRAY: Boolean false => return 0', () => {
  expect(TSarray.format(null)).toBe('0');
});

// TEST WITH VALUES
test('ARRAY: Value is lorem123 => return 0', () => {
  expect(TSarray.format('lorem123')).toBe('0');
});
test('ARRAY: Value lenght is > 15 => return 0', () => {
  expect(() => {
    TSarray.format('012345678901010101');
  }).toThrow();
});
